src/java/org/apache/cassandra/db/rows/Cell.java

```diff
@@ -59,14 +68,24 @@
 
     public interface Factory<V>
     {
-        Cell<V> create(ColumnMetadata column, long timestamp, int ttl, int localDeletionTime, V value, CellPath path);
+        Cell<V> create(ColumnMetadata column, long timestamp, int ttl, long localDeletionTime, V value, CellPath path);
     }
 
     protected Cell(ColumnMetadata column)
     {
         super(column);
     }
 
+    public static int deletionTimeLongToGuavaUint(long deletionTime)
+    {
+        return deletionTime == NO_DELETION_TIME ? UnsignedInteger.MAX_VALUE.intValue() : UnsignedInteger.valueOf(deletionTime).intValue();
+    }
+
+    public static long deletionTimeGuavaUintToLong(int deletionTimeGuavaUint)
+    {
+        return deletionTimeGuavaUint == UnsignedInteger.MAX_VALUE.intValue() ? NO_DELETION_TIME : UnsignedInteger.fromIntBits(deletionTimeGuavaUint).longValue();
```

Comment:
Instead of `UnsignedInteger.fromIntBits(deletionTimeGuavaUint).longValue()`, perhaps we could use `UnsignedInts.toLong(deletionTimeGuavaUint)` and skip the instantiation of `UnsignedInteger`.

That method is marked as beta, which is always scary. However, `UnsignedInteger#longValue` still calls to it, so we are using the beta method anyway.

Reply:
Why not just `& 0xffffffffL`?

Reply:
`UnsignedInts` is sthg I started with. You can see in the trail of commits it was there. I have a very strong opinionated opposition to using it. This comes from the experience of working in this PR and the diabolical debug sessions I had to deal with.

Think that `UnsignedInts` will just swallow any value you throw at it. It doesn't sanity check anything. The bugs that emerge are just gibberish timestamps that don't _crash_ the code. You just get silent row removal, silent row mismatch, etc. Just think in terms of the code randomly changing timestamps and how this would fan out.

Given the above consider also the future of the codebase. If we sometime build some new code that accidentally produces an invalid value the bugs that will emerge from that will be probably silent and 100% sure hard to find. Think sentinel values, usually 'out of band' values that can easily fall outside a valid uint range...

It would be very painful for me to move away from `UnsignedInteger` knowing what I've learnt, it would be very unwise imo and an ugly accident waiting to happen. I would strongly recommend staying with 'UnsignedInteger'.

Reply:
What is it you think you’re gaining with this boxing? I have looked, and I cannot see any additional protection, and we have to rely on escape analysis to avoid significant performance penalty. 

This kind of bit twiddling is all over the codebase and we can manage it very successfully. It *shouldn’t* be seen as exotic or dangerous.

Reply:
And I totally agree bitwise operations are not dangerous or exotic. But I'd rather run hot needles through my eyes than removing that and have to deal with such bugs in the future I've had to deal with when it wasn't present. 

It does provide protection which has proven invaluable.
```
java.lang.IllegalArgumentException: value (-207332) is outside the range for an unsigned integer value
[junit-timeout] 	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:202)
[junit-timeout] 	at com.google.common.primitives.UnsignedInteger.valueOf(UnsignedInteger.java:75)
[junit-timeout] 	at org.apache.cassandra.db.rows.Cell.deletionTimeLongToGuavaUint(Cell.java:81)
```
It's an accident waiting to happen, we're shooting ourselves in the foot imo.

Reply:
So you're talking about the reverse conversion. The annotated code is not protected at all, and so there is no reason to box the value here. But, yes, this protection can be easily reproduced with a static method for converting from long to int, that confirms the upper bits are unset, i.e. simply checks (value & 0xffffffff00000000L) == 0.

There is _no_ reason to use a boxing method for this.

Reply:
Agree, `UnsignedInteger` provides validation for the long -> uint conversion, but not for the uint -> long conversion discussed here. `UnsignedInteger.fromIntBits(deletionTimeGuavaUint).longValue()` doesn't do any validation, and it's equivalent to `UnsignedInts.toLong(deletionTimeGuavaUint)` or `deletionTimeGuavaUint & 0xffffffffL`.

Reply:
Yep, and we should remove its usage in all places and replace it with a suitable static method that performs whatever sanity checks we want without boxing.

Reply:
Right you are, icym now. Fixed.